scripts/completions/zsh-completions/_wl-mirror

```diff
@@ -0,0 +1,73 @@
+#compdef wl-mirror
+
+_wl_mirror_outputs() {
+    swaymsg -t get_outputs | sed -n '/^[[:space:]]*"name":/s/^.*"\([^"]*\)",/\1/p'
```

Comment:
Completing outputs would be cool, but this implementation would only work on sway. There sadly isn't a simple portable way to get a list of outputs without a dedicated tool.

One solution in the long term could be a family of extra options to the main wl-mirror binary to assist scripts in information-gathering (e.g. something like `--print-outputs`) that print some information and then exit.

Reply:
Right, I didn't think about non-swaymsg based systems.

I have my own `wl-*` scripts for similar purposes (i.e. [wl-scripts](https://github.com/mazunki/wl-scripts)), but I currently rely on `swaymsg` for most of it. Is there some standardized way to access this information at all through Wayland? Maybe I should extend my scripts for this... hm.

Reply:
every wayland client receives a list of global objects upon connecting. one of the object kinds received is a list of all current outputs (with events when an output is added/removed).

The simplest solution is probably a small CLI utility dedicated to performing such one-off queries (connect, collect info like outputs, print, exit). Since that part of wayland is standardized, it will work on every compositor.

There currently is no widely used CLI tool like that that I know of.

Reply:
as a quick non-exhaustive list of what info is easily accessible:

- outputs (via `wl_output` global and `xdg_output` extension)
  - name/connector (e.g. `eDP-1`), manufacturer, model, resolution (pixels), resolution (logical/after scaling), position (logical)
- toplevels / windows (via `ext_foreign-toplevel_list` and/or compositor-specific extensions like `wlr_foreign_toplevel_management`)
  - title, app id, identifier
  - windows position is not part of the standard protocol, but IIUC the wlroots version exposes window position
- ...